File: Elixir/3_merge_two_sorted_lists.ex
# Definition for singly-linked list.
#
# defmodule ListNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           next: ListNode.t() | nil
#         }
#   defstruct val: 0, next: nil
# end

defmodule Solution do
  @spec merge_two_lists(list1 :: ListNode.t | nil, list2 :: ListNode.t | nil) :: ListNode.t | nil
  def merge_two_lists(list1, list2) do
    cond do
      is_nil(list1) -> list2
      is_nil(list2) -> list1
      list1.val < list2.val ->
        %ListNode{val: list1.val, next: merge_two_lists(list1.next, list2)}
      true ->
        %ListNode{val: list2.val, next: merge_two_lists(list1, list2.next)}
    end

File: Erlang/3_merge_two_sorted_lists.erl
%% Definition for singly-linked list.
%%
%% -record(list_node, {val = 0 :: integer(),
%%                     next = null :: 'null' | #list_node{}}).

-spec merge_two_lists(List1 :: #list_node{} | null, List2 :: #list_node{} | null) -> #list_node{} | null.
merge_two_lists(List1, List2) ->
    case {List1, List2} of
        {null, _} -> List2;
        {_, null} -> List1;
        {#list_node{val = V1, next = N1}, #list_node{val = V2, next = N2}} when V1 < V2 ->
            #list_node{val = V1, next = merge_two_lists(N1, List2)};
        {L1, #list_node{val = V2, next = N2}} ->
            #list_node{val = V2, next = merge_two_lists(L1, N2)}
    end.